Src/IronPython/Compiler/Parser.cs

```diff
@@ -1737,12 +1766,18 @@ struct WithItem {
                     return ErrorStmt();
                 }
 
-                if (metaclass != null) {
-                    l.Add(new AssignmentStatement(new[] { new NameExpression("__metaclass__") }, metaclass));
+                if (clsContext != null) {
+                    if (!clsContext.InMember && clsContext.Metaclass != null) {
+                        l.Add(new AssignmentStatement(new[] { new NameExpression("__metaclass__") }, clsContext.Metaclass));
+                    }
+                    
+                    if (clsContext.InMember) {
+                        l.Add(new AssignmentStatement(new[] { new NameExpression("__class__") }, new NameExpression(clsContext.Name)));
```

Comment:
While this would work as an initial implementation but I don't think it should be done at this level. I think I have something that might work for this, if I have time I'll post something this afternoon...

The tests below pass in CPython but fail with this PR:
```Python
class test:
  def test(self):
    assert "__class__" not in locals()

test().test()

class test:
  def test(self, __class__=1):
    assert __class__ == 1

test().test()
```